### ISSUE TITLE
The BasisEmbedding template does not allow for all-ones or all-zero basis states

### ISSUE THREAD

**Issue by josh146** [PennyLaneAI/pennylane#1112]
There is a bug in the input validation for the `BasisEmbedding` template. It should be checking that the basis state array only contains 0 and 1 integers, but the way it is currently coded, it also disallows all-ones and all-zero basis states:

https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/templates/embeddings/basis.py#L54

**Comment by antalszava** [comment]
Indeed, the following raises an error when using `BasisEmbedding`:

```python
import pennylane as qml
import numpy as np

dev = qml.device('default.qubit', wires=[0,1,2])

@qml.qnode(dev)
def circuit():
    qml.templates.embeddings.BasisEmbedding(np.array([1,1,1]), wires=[0,1,2])
    return qml.expval(qml.PauliZ(0))

circuit()
```
```
ValueError: Basis state must only consist of 0s and 1s; got [1, 1, 1]
```
However, this got me wondering: isn't `BasisEmbedding` identical to[ `BasisStatePreparation`](https://github.com/PennyLaneAI/pennylane/blob/5095918c03da1b82bbdbe94055d7846244127dc0/pennylane/templates/state_preparations/basis.py#L76)? 🤔

For the latter, the check is correct and it seems as though it had the same functionality. Maybe @mariaschuld could also help here. If they are identical, we could maybe consider having `BasisEmbedding` depend on `BasisStatePreparation` and eventually deprecating it.

**Comment by mariaschuld** [comment]
Yes, there is no real reason why we couldn't use `BasisStatePreparation`. I think I always thought it hacks the initial state, while I wanted a hardware-compatible solution for the embedding, but I guess this argument is not valid (any more). But this is a general decision, I propose to decouple it from this bug.

I made a fix in PR #1114...
